relayer/workers/parachaincommitmentrelayer/ethereum-channel-writer.go

```diff
@@ -0,0 +1,137 @@
+package parachaincommitmentrelayer
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/sirupsen/logrus"
+
+	"github.com/snowfork/polkadot-ethereum/relayer/chain"
+	"github.com/snowfork/polkadot-ethereum/relayer/chain/ethereum"
+	"github.com/snowfork/polkadot-ethereum/relayer/contracts/inbound"
+	"github.com/snowfork/polkadot-ethereum/relayer/substrate"
+)
+
+type EthereumChannelWriter struct {
+	config    *ethereum.Config
+	conn      *ethereum.Connection
+	contracts map[substrate.ChannelID]*inbound.Contract
+	messages  <-chan []chain.Message
+	log       *logrus.Entry
+}
+
+func NewEthereumChannelWriter(config *ethereum.Config, conn *ethereum.Connection, messages <-chan []chain.Message,
+	contracts map[substrate.ChannelID]*inbound.Contract,
+	log *logrus.Entry) (*EthereumChannelWriter, error) {
+	return &EthereumChannelWriter{
+		config:    config,
+		conn:      conn,
+		contracts: contracts,
+		messages:  messages,
+		log:       log,
+	}, nil
+}
+
+func (wr *EthereumChannelWriter) Start(ctx context.Context, eg *errgroup.Group) error {
+
+	id := substrate.ChannelID{IsBasic: true}
+	contract, err := inbound.NewContract(common.HexToAddress(wr.config.Channels.Basic.Inbound), wr.conn.GetClient())
+	if err != nil {
+		return err
+	}
+	wr.contracts[id] = contract
+
+	id = substrate.ChannelID{IsIncentivized: true}
+	contract, err = inbound.NewContract(common.HexToAddress(wr.config.Channels.Incentivized.Inbound), wr.conn.GetClient())
+	if err != nil {
+		return err
+	}
+	wr.contracts[id] = contract
+
+	eg.Go(func() error {
+		return wr.writeMessagesLoop(ctx)
+	})
+
+	return nil
+}
+
+func (wr *EthereumChannelWriter) onDone(ctx context.Context) error {
+	wr.log.Info("Shutting down writer...")
+	// Avoid deadlock if a listener is still trying to send to a channel
+	for range wr.messages {
+		wr.log.Debug("Discarded message")
```

Comment:
Is there a call missing on each message? We are just logging here?

Reply:
could be yeh - this code was just ported over from old code. will fix deadlock/restart logic in upcoming PR